Phone and tablet UX + mmcboot support

samueldr opened the pull request:
* * *

### *Give me the build*

Woah there, you probably want to read and review before, but eh, if you want it, here it is:

#### Build 004.5

**Built against e65e42de0817bf3c4baf9ef73d2bc33db5911d30**

| Build | Installation notes | Troubleshooting guide |
|---|---|---|
| [build 004.5 Pinephone A64 build](https://stuff.samueldr.com/7pnblffpscrmx51vlbl1vky53rizhxqv-Tow-Boot.release.004-pre/pine64-pinephoneA64-2021.10-004-pre.tar.xz) | [Installation notes](https://github.com/Tow-Boot/Tow-Boot/blob/e65e42de0817bf3c4baf9ef73d2bc33db5911d30/boards/pine64-pinephoneA64/INSTALLING.md) | *(none)* |
| [build 004.5 Pinephone Pro build](https://stuff.samueldr.com/7pnblffpscrmx51vlbl1vky53rizhxqv-Tow-Boot.release.004-pre/pine64-pinephonePro-2021.10-004-pre.tar.xz) | [Installation notes](https://github.com/Tow-Boot/Tow-Boot/blob/e65e42de0817bf3c4baf9ef73d2bc33db5911d30/boards/pine64-pinephonePro/INSTALLING.md) | [Troubleshooting guide](https://gist.github.com/samueldr/96171ff1d85b682b067753b1cf57f9e1) |

<details><summary><b>Outdated builds</b></summary>

 - ~~[WIP 1 Pinephone Pro build](https://stuff.samueldr.com/Tow-Boot/pine64-pinephonePro-2021.10-003-pre.tar.xz)~~
 - ~~[WIP 2 Pinephone Pro build](http://stuff.samueldr.com/8rqqjxr95ck6n9yaysw1r2jqpbzi3nbk-pine64-pinephonePro-2021.10-004-pre.tar.xz)~~

 - ~~Built against db86fbfd341e7f5366217bea91cfb01a948ce54a~~
     - ~~[WIP 3 Pinephone A64 build](http://stuff.samueldr.com/h4d0a6sj7jr425vpgdprk6d7bab8cy72-pine64-pinephoneA64-2021.10-004-pre.tar.xz)~~
     - ~~[WIP 3 Pinephone Pro build](http://stuff.samueldr.com/j8ylph2377b4z3rp0m4j879h7lp6l3r3-pine64-pinephonePro-2021.10-004-pre.tar.xz)~~

 - ~~Built against ccf5627f5977ebf9637d6ac495c63690f2145223; main difference: mmcboot installer for Pinephone A64.~~
     - ~~[WIP 4 Pinephone A64 build](http://stuff.samueldr.com/fdb270zzddxh41a7dyv8vgck6f3svmva-pine64-pinephoneA64-2021.10-004-pre.tar.xz)~~
     - ~~[WIP 4 Pinephone Pro build](http://stuff.samueldr.com/pfyi82x1x7zckxqbywpf31p5l49pfiij-pine64-pinephonePro-2021.10-004-pre.tar.xz)~~

</details>

* * *

### TODO

 - [x] Final reviews / approval by peers

* * *

### What is this?

This is my attempt at making boot on "Linux phones" as painless as possible.

Features:

 - LED management
 - Phone keys based control
 - User-friendlier universal defaults
 - "Status reporting"

Anti-features:

 - Superficial status reporting without display enablement.

#### LED management

The LED is lit-up as early as possible, assuming it's RGB, it's lit-up red first.

When Tow-Boot starts running the default distro boot (as defined by U-Boot), it turns yellow.

Some additional boot modes will turn the LED different colours:

 - Blue for Mass Storage
 - Aqua for *requesting* SD card boot

#### Phone keys based control

As commonly done with "consumer" phones and mobile devices, holding volume keys during early boot phases will affect the boot in different manner.

 - Volume up: Boot in USB Mass Storage mode
 - Volume down: Request Operating System Boot from SD card

> **Note**: *USB Mass Storage* (or UMS) exports the eMMC (internal) of the device over USB.

#### User-friendlier universal defaults

To make things less confusing across devices, **all** devices will load the operating system using the same rules.

The schemes supported by the distro boot scheme from U-Boot are all tried on the following devices in order:

 - eMMC
 - SD

The order is static and *eMMC first* **on all devices**. This is a deviation from the defaults from the different communities, and somewhat a deviation from the defaults of U-Boot. Doing so helps reduce confusion and should help reduce support load.

This helps reduce confusion, because the order the operating systems will be tried to be loaded is the same, wherever Tow-Boot was started from. This is quite important considering different SoCs have a different BootROM startup order, making it harder to "just" boot the SD card if started from the SD card.

Furthermore, attempting Operating System boot from eMMC first is a done considering *booting from a removable storage should be a conscious choice, **requested** by the end-user*.

> **Note**: When all Operating System boot sources have been attempted, and none worked, the device will poweroff after reporting the failure.

Not specific to the Phone UX, Tow-Boot *always uses 115200 baud rate*. This way, the UART rate is always the same, helping a bit more with unified instructions.

#### "Status reporting"

In addition to the LEDs reporting the current boot state, vibration support has been added. The device will vibrate once as early as possible.

When failing to do the Operating System boot from SD *as requested by the user* (holding volume down), the LED will flash 4 times, and the device will vibrate accordingly. Boot will continue to the default distro boot.

When failing to do the default distro boot, the LED will flash 10 times, and the device will vibrate accordingly. Then the device will poweroff.

> **Note**: when connected to a power source, some devices (the Pinephone Pro) will not poweroff, and will instead reboot.

In addition to that, with the shared disk image, a *protective partition* is added to the GPT to prevent bad manipulations from breaking the installed *Platform Firmware*. This protective partition also provides strong guarantees when updating the Tow-Boot install, such that when its size increases, it is assured not to interfere with the other partitions.

#### Superficial status reporting without display enablement

As described previously, all of this is done such that the user knows what is happening, even though the display shows nothing.

This, to me is a sort of "anti-feature", but in reality we must keep this scheme such that further devices can use it even if a display driver has not been written yet for them.

* * *

### How do I install this?

Ideally this would be installed to SPI flash using an SPI flash installer. An implementation was written with a "specialized" Linux system build so that the initial lack of display enablement on the Pinephone Pro wouldn't be an issue.

Alas, none of the targeted devices support SPI boot, thus none can be installed to dedicated storage.

To install to eMMC, the steps will generally be the same:

 - **These instructions will clobber the content of the SD card and the eMMC, defining a new partition table.**
 - Write the `shared.disk-image.img` to an SD card as-is
 - Boot the device to USB Mass Storage by holding Volume Up
 - Write the `shared.disk-image.img` to the phone.

> **Note**: With some SoCs (e.g. Rockchip) you will need to ["neuter" the *Platform Firmware*](https://github.com/Tow-Boot/Tow-Boot/issues/47) installed to the eMMC if there is one already installed. 

### How do I install a Linux distro after installing Tow-Boot?

This is a hard question to answer correctly. This will require a change in habits of the different distributions.

<details><summary>Misc. distro notes</summary>

#### Mobile NixOS

I can only answer for the distribution I'm involved with, Mobile NixOS.

A temporary installation method has been added:

 - https://github.com/NixOS/mobile-nixos/pull/442
 - https://github.com/NixOS/mobile-nixos/pull/445

It relies on a script and Mass Storage support on the target device. This is a method I think would be acceptable to keep recommending for other distros. The main thing to understand here is the script manipulates the partitions, then splats the content in them.

> The permanent solution will be an *installer system*. The user will write Tow-Boot to the eMMC, and boot from a *universal* SD card that holds an installer system that will guide the user through some questions about partitioning (e.g. ask about FDE and filesystems), the preferred graphical environment, and then do a "classic" opinionated `nixos-install` under the covers.

</details>

* * *

Handles part of #104. (Devices listing on the website.)

Danct12 commented:
For mass storage mode, is it possible to export the SD card as well? It would be nice if you can write to SD card from there.

samueldr commented:
Unless I missed something when I looked, the UMS driver for U-Boot *at this point in time* only handles exporting one LUN.

With UX improvements (mainly display output support) we could let the user pick which storage medium is exported.

An alternative could be to look into adding support for exporting more LUNs in upstream U-Boot.

samueldr committed:
tow-boot/phone-ux: Port from the WIP branch

This is much easier to deal with, compared to the WIP branch's
implementation.

Note that this probably requires some rework still, for now we're
porting it 1:1.

samueldr committed:
pine64-pinephoneA64: Init with phone-ux

samueldr committed:
pine64-pinephonePro: Init with phone-ux

A direct port from the initial porting branch.

samueldr committed:
pine64-pinephonePro: Add SPI support

samueldr committed:
pine64-pinephonePro: Finish implementing SPI support

Be-ing commented:
> An alternative could be to look into adding support for exporting more LUNs in upstream U-Boot.

I think letting the user choose which device is exported would be more foolproof. With both the SD and eMMC exported, the user could accidentally write to the wrong device. With Jump Drive on the PinePhone, I've had to check the drive size reported in dmesg to ensure I'm writing to the correct device. In the event the user is using an SD card that happens to be the same size as the eMMC drive, that could be tricky.

samueldr committed:
ELO: Add pine64-pinephonePro

samueldr committed:
ELO: Import first draft installer app

samueldr committed:
modules/build: Provide access to the SPI eval

samueldr committed:
tow-boot/installer: Directly use `firmwareSPI`

The other method is going to cause issues with the upcoming feature...

samueldr committed:
ELO: Ignore result

samueldr committed:
tow-boot/installer: Add "touch" installer

samueldr committed:
ELO: pin celun

samueldr commented:
This push rebases the work that was on this branch, with the refreshed work built on top of #78.

That rewrite was done **because** this PR showed it was really needed (in addition to making other things easier).

This push also includes the "ELO" changes, "Embedded Linux OS", which is used to build a touch-based installer system for installing to SPI, for devices that support installation to SPI.

samueldr committed:
pine64-pinephonePro: Update base support patch

Remove SPI support from it (except for the definition in the DT)

samueldr committed:
pine64-pinephonePro: Add back SPI support changes

This time as a discrete patch

samueldr committed:
pine64-pinephonePro Work around eMMC issues with UMS

It seems the current patch series works fine for the tested scenarios,
including the ones I had previously found out.

But it's not enough for UMS to work like it did previously.

The following messages will be observed:

```
*** ERROR: Can't read GPT header ***
find_valid_gpt: *** ERROR: Invalid GPT ***
mmc fail to send stop cmd
```

Though using the configuration just added seems to work around the
issues, unclear why. These were the options the patch author was testing
against.

This could come from insufficient backporting. An alternative would be
to go back to the previous workaround, and disable the feature that is
the root cause of issues.

samueldr committed:
pine64-pinephoneA64: Drop vibrator patch

samueldr committed:
pine64-pinephonePro: Drop vibrator patch

samueldr committed:
modules/phone-ux: Add vibrator patch

samueldr committed:
pine64-pinephonePro: reorganize patches

First, generic non-device-specific changes

Then, the board enablement.

Strit commented:
Just writing to confirm that the WIP2 builds work fine on my Pinephone Pro.
It even makes the phone able to suspend!

Does the WIP3 build add something or is it just re-arranging some stuff?

samueldr commented:
Re-arranging stuff, *adding a Pinephone A64 pre-built artifact*, and fixes the OS boot order regression.

But otherwise mostly the same.

samueldr committed:
modules: Add basic requirements for mmcboot variant

samueldr committed:
tow-boot/installer: Implement mmcboot "harness" around installer

The actual install script needs to be written.

a-wai commented on boards/pine64-pinephoneA64/default.nix:
```diff
@@ -0,0 +1,56 @@
+      (helpers: with helpers; {
+        USB_MUSB_GADGET = yes;
+        USB_GADGET_MANUFACTURER = freeform ''"Pine64"'';
+      })
```
```suggestion
      })
      (helpers: with helpers; {
        SUPPORT_EMMC_BOOT = yes;
      })
```

This option would allow u-boot to boot from eMMC boot partitions on the A64.

a-wai replied:
Also, upstream u-boot has support for startup LED (since 2022.01 IIRC) in case you hadn't noticed, so an additional patch isn't needed for that (vibrator isn't supported upstream, however).

(I have to admit I didn't look up what u-boot version this PR was based on though, so this comment might or might not be relevant)

samueldr replied:
The patch implements LED startup like other RK3399 boards implement it.

I seem to remember looking at implementing it "correctly" through their framework, and that it could not be used during SPL since it brings in too much code to fit.

samueldr replied:
> This option would allow u-boot to boot from eMMC boot partitions on the A64.

Stay tuned for the A64, work is in progress for support for eMMC boot partitions.

samueldr replied:
Oh, forgot to come back to this discussion and say it: it's done. It works. Tested against Allwinner and Amlogic targets.

samueldr committed:
radxa-zero2: Build an mmcboot image

samueldr committed:
tow-boot/installer: Add mmcboot installer script

samueldr committed:
hardware/amlogic: Fix mmcboot binary for G12

The mmcboot binary will be pre-offset to LBA1, as required for writing
to an mmcboot partition.

We prefer writing the binary file from 0x00 all the time, making things
easier to implement.

jussihi commented:
Hello, the download links don't seem to work for me for some reason. Is there some problems with the domain name redirection?

samueldr commented:
AFAICT it still is working just fine here. `wget` or `curl` output might help track down issues.

jussihi commented:
> AFAICT it still is working just fine here. `wget` or `curl` output might help track down issues.

Thank you, it actually worked, it was just a bug or something in my web browser. I was able to download by copying the link to another tab (?!). Thanks for the work, now I'm rocking Tow-Boot on my PPP! It works great!

samueldr committed:
orangePi-zeroPlus2H5: Build an mmcboot image

samueldr committed:
pine64-pinebookA64: Build an mmcboot image

justuser-ppp commented:
Hello all. I erased spi and flashed it. Now I cannot boot from sdcard. Is there any way to recover from this? Thank you for your help.

justuser-ppp commented:
Apparently I still can boot from sdcard if i continue spamming volume down button on boot. Everything works as it should after. Strange.

PureTryOut commented:
Note that you don't need to erase the SPI before flashing it, the flash button already handles that. Also please don't ask for user support here but join [#tow-boot:matrix.org](https://matrix.to/#/#tow-boot:matrix.org) instead.

justuser-ppp commented:
Thank you. Joined;)

samueldr committed:
hardware/allwinner: Add commands for mmcboot

samueldr committed:
tow-boot/installer: Make touch installer target block configurable

This was not much of an issue with mtdblock0... But with the eMMC boot
partition installer, we need to know what to target!

samueldr committed:
ELO: Add emmcboot support to touch installer config

samueldr committed:
tow-boot/installer: Add emmcboot touch-installer support

samueldr committed:
pine64-pinephoneA64: Add mmcboot support

samueldr committed:
ELO: Add pine64-pinephoneA64 support

samueldr committed:
ELO: Ensures builds include gadget support

This may need to be reviewed once ELO is used for more than the phone
UX.

samueldr committed:
ELO: Implement allwinner mmcboot in touch installer

samueldr committed:
ELO: Update touch installer

samueldr commented:
It was beginning to be too cumbersome to separate the changes for mmcboot support, when this PR touched the installer config.

I hate bundling more features into a huge PR, but this had to be done.

samueldr committed:
overlay: Add crustFirmware builder

Co-authored-by: Daniel Barlow <dan@telent.net>

samueldr committed:
hardware/allwinner: Build crust SCP firmware

samueldr committed:
orangePi-zeroPlus2H5: Configure crust defconfig

samueldr committed:
olimex-teresI: Disable crust build

samueldr committed:
tow-boot/installer: Fix path reference

This accidentally used a new syntax(?) that works with Nix 2.4

chookity-pokk commented:
> Note that you don't need to erase the SPI before flashing it, the flash button already handles that. Also please don't ask for user support here but join [#tow-boot:matrix.org](https://matrix.to/#/#tow-boot:matrix.org) instead.

The tow-boot matrix doesn't seem to be join-able anymore?

samueldr commented:
That's likely because to discover rooms, sometimes "via" servers are needed, try:

 - https://matrix.to/#/#Tow-Boot:matrix.org?via=matrix.org

samueldr committed:
u-boot: Drop workaround for env from boot order patch

The workaround is replaced by an appropriate fix.

samueldr committed:
hardware/allwinner: Fix build with environment outside mmc

samueldr committed:
tow-boot: Add (private) variant for installer usage

This variant will make the SD card preferred over internal storage.

samueldr committed:
tow-boot: Enable ordering patches according to variant

Note that the boot-installer variant is private and made exclusively for
the installers.

samueldr committed:
pine64-pinebookA64: Fix device name

samueldr committed:
modules/metadata: Init module to output metadata about a device

samueldr committed:
boards/radxa-RockPi4: Fix eval

samueldr committed:
boards/raspberryPi-aarch64: White lie waiting for the change

This is needed since it's part of the metadata.

samueldr committed:
support/release-tools: Handle released devices

samueldr committed:
release.nix: Use release-tools.releasedDevicesEvaluations

samueldr committed:
doc: Add automatic devices list

This is translated to markdown from the Mobile NixOS implementation.

samueldr committed:
doc: Apply proper styles for device sidebar

samueldr committed:
doc: Link devices in header

samueldr committed:
doc: Include device pages

samueldr committed:
default.nix: Use release-tools

samueldr committed:
doc: Generate listing grouped by manufacturer

samueldr committed:
doc: Update sidebar

samueldr committed:
modules/documentation: Init module with documentation fragments

samueldr committed:
modules/metadata: Provide documentation fragments

samueldr committed:
raspberryPi-aarch64: Add installation instructions

samueldr committed:
hardware/allwinner: Add installation instructions...

... except for the phone UX.

samueldr committed:
doc: Add installation instructions to device pages

samueldr committed:
hardware/amlogic: Add default installation steps

samueldr committed:
hardware/rockchip: Add default installation steps

samueldr committed:
boards/odroid-C2: Add default installation steps

Reminder: this is a bespoke build process.

samueldr committed:
boards/libreComputer-amlS805xAc: Add installation notes

This board is slighly special since it doesn't have a trivial to use
"external storage" that is bootable.

samueldr committed:
boards/pine64-pinephoneA64: Add device and installation notes

samueldr committed:
boards/pine64-pinephonePro: Add device and installation notes

samueldr committed:
doc: Fix image names

samueldr commented on boards/pine64-pinebookA64/default.nix:
```diff
@@ -1,12 +1,13 @@
   device = {
     manufacturer = "PINE64";
-    name = "A64-LTS";
+    name = "Pinebook (A64)";
```
shhhh

ItsHarper commented on boards/pine64-pinephoneA64/README.md:
```diff
@@ -0,0 +1,16 @@
+blue if done successfully. In this mode, the phone will work like a USB drive
+when connected to a host computer.
+
+Booting an operating system, or an installer written to an SD card will require
+holding the *volume down* button before and during the second vibration. When
+done successfully, the LED will turn aqua.
```
```suggestion
Booting from an SD card will require holding the *volume down* button before and
during the second vibration. When done successfully, the LED will turn aqua.
```
The way you had it worded could be interpreted as saying "Booting an operating system will require holding down the volume down button".

samueldr replied:
I edited the wording to include *operating system* still, as I try very hard in the documentation to strongly delineate between *platform firmware startup* and *operating system boot*. This is because the former is generally static and set in hardware, while the latter is handled by the platform firmware, and can be anything.

ItsHarper commented on boards/pine64-pinephonePro/README.md:
```diff
@@ -0,0 +1,16 @@
+blue if done successfully. In this mode, the phone will work like a USB drive
+when connected to a host computer.
+
+Booting an operating system, or an installer written to an SD card will require
+holding the *volume down* button before and during the second vibration. When
+done successfully, the LED will turn aqua.
```
```suggestion
Booting from an SD card will require holding the *volume down* button before and
during the second vibration. When done successfully, the LED will turn aqua.
```

ItsHarper commented:
I just installed this onto my A64 Pinephone. The installation process was very smooth, and thanks to the new mmcboot support, it didn't wipe my existing OS installation! Awesome work.

ItsHarper commented:
This PR should get linked to #76

samueldr committed:
pine64-pinephoneA64: reword external boot

Co-authored-by: Noah Andrews <noah@noahandrews.me>

samueldr committed:
pine64-pinephonePro: reword external boot

Co-authored-by: Noah Andrews <noah@noahandrews.me>
